### ISSUE TITLE
Increasing # of observations scales linearly with time usage

### ISSUE THREAD

**Issue by PiranjaF** [tensorflow/serving#322]
Hi,

I'm running an LSTM model on a CPU-based Linux server with the newest version of TF Serving. Strangely enough, running e.g. 'N=10' observations takes 10 times as long as running 'N=1' observations in the model. Given that LSTM models are implemented using matrix multiplications the time usage should not scale linearly. I run the model using the following code, where 'npy' is a numpy array of shape '(N, 30)'.  

```
request.inputs['input_1:0'].CopyFrom(tf.contrib.util.make_tensor_proto(npa.astype('int32')))
```

Any insights into what I might  be doing wrong?

**Comment by kirilg** [comment]
What scaling were you expecting? I think roughly linear makes sense unless it can take advantage of some parallelism. Is N the number of steps in your LSTM or the number of input examples?

**Comment by PiranjaF** [comment]
N is the number of input examples. Given that everything in the LSTM is matrix multiplications, I was definitely expecting better than linear scaling. For instance, during training, doing a forward-pass with a batch size of 512 is not 8 times slower than with a batch size of 64. In my experience, it's more like 4 times slower.

**Comment by kirilg** [comment]
Interesting. Just to confirm since you said you're using CPU for serving, were you using CPU for training as well when seeing this sub-linear scaling (i.e. not using GPU)?

I'm probably missing something, but at least for smaller non-square matrices, doesn't matrix multiplication scale linearly in the batch dimension? (unless of course you can take advantage of some parallelism on GPUs for example). Above a certain size threshold, the underlying libraries may get you better scaling in the batch dimension. For example, in training you mentioned 512 vs. 64 (input matrix can be large enough to start taking advantage of optimizations) but at serving time you're comparing 10 vs. 1. Can you try using the same larger batch sizes at serving and compare?

**Comment by gautamvasudevan** [comment]
Closing due to staleness. If this is still an issue, please file a new updated issue with current steps to reproduce the bug. If this is a question, please ask it on: 

https://stackoverflow.com/questions/tagged/tensorflow-serving

Thanks!
